simple/util/file_match.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+
+import fs.path as fspath
+from util.filesystem import File
+
+
+def match(f: File, pattern: str) -> bool:
+  """Returns true if this file's name or path matches a given pattern.
+
+    Pattern rules:
+    - A leading double slash means the path must match from the root of the input dir.
+    - Protocol (e.g. gs://) can optionally be included in the pattern and will also add the requirement of matching from the root.
+    - A single wildcard (*) in the name or protocol portions of a pattern can be zero or more characters.
+      - Double wildcards for name or protocol are invalid; use single wildcards instead.
+    - A single wildcard (*) in the directory portion of a pattern can be any non-slash character.
+    - A double wildcard (**) in the directory portion of a pattern can be any character including slashes, i.e. multiple dir levels.
+
+    Since matching is ultimately done with regular expressions, it may be
+    possible to do more complex matching by passing in patterns containing special
+    characters, but this behavior is not officially supported and may break at any time.
+
+    Examples:
+    - Match any CSV file in any directory or subdirectory: "*.csv"
```

Comment:
Will this change the current default behavior where it does not look for files in sub directories? If it does, it may pick up files that were not previously picked.

On the larger point of sub-directories, are you thinking it will be an opt-in (via a config field) or will that be the default moving forward?

Reply:
Yes, this would change that default behavior. For that reason, we would also need to stop allowing output dir to overlap with input dir. I imagined this would be the default moving forward.